src/devices/bus/mca/ibm72x8299.cpp

```diff
@@ -0,0 +1,387 @@
+		case 1: return "ADAPTER_ID_HI";
+		case 2: return "OPTION_SELECT_DATA_1";
+		case 3: return "OPTION_SELECT_DATA_2";
+		case 4: return "OPTION_SELECT_DATA_3";
+		case 5: return "OPTION_SELECT_DATA_4";
+		case 6: return "SUBADDRESS_EXT_LO";
+		case 7:	return "SUBADDRESS_EXT_HI";
+		default: return "(error)";
+	}
+}
+
+uint8_t ibm72x8299_device::pos_registers_r(offs_t offset)
+{
+	// Programmable Option Select registers.
+	// 100: adapter identification low
+	// 101: adapter identification high
+	// 102: option select data 1
+	// 103: option select data 2
+	// 104: option select data 3
+	// 105: option select data 4
+	// 106: subaddress extension low
+	// 107: subaddress extension high
+
+	// Special case: Reading the planar VGA's registers.
+	if(BIT(m_system_board_setup, 5) == 0)
+	{
+		if(!machine().side_effects_disabled()) LOGPOS("%s: reading VGA POS register %s\n", FUNCNAME, posreg_name(offset));
+
+		// VGA I/O byte.
+		if(offset == MCABus::POS::OPTION_SELECT_DATA_1)
+			return dynamic_cast<mca16_planar_vga_device *>(m_planar_vga->get_card_device())->sleep_r();
+		else return 0xFF;
+	}
+	// Special case: Reading the planar's registers rather than an adapter.
+	if(!BIT(m_system_board_setup, 7))
+	{
+		if(!machine().side_effects_disabled()) LOGPOS("%s: reading planar POS register %s\n", FUNCNAME, posreg_name(offset));
+		return m_planar->planar_pos_r(offset);
+	}
+
+	if(m_planar->get_mca_slot(m_adapter_setup & 7) != NULL)
+	{
+		auto card = dynamic_cast<device_mca16_card_interface *>(m_planar->get_mca_slot(m_adapter_setup & 7)->get_card_device());
+		if(card)
+		{
+			uint8_t value = (card ? card->pos_r(offset) : 0xFF);
+			LOGPOS("%s: reading adapter %d POS register %s: %02X\n", FUNCNAME, (m_adapter_setup & 7) + 1, posreg_name(offset), value);
+			return value;
+		}
+		else
+		{
+			LOGPOS("%s: no card in slot %d\n", FUNCNAME, (m_adapter_setup & 7) + 1);
+			return 0xFF;
+		}
+	}
+	else return 0xFF;
+}
+
+void ibm72x8299_device::pos_registers_w(offs_t offset, uint8_t data)
+{
+
+	if(BIT(m_system_board_setup, 5) == 0)
+	{
+		if(!machine().side_effects_disabled()) LOGPOS("%s: writing VGA POS register %s: %02X\n", FUNCNAME, posreg_name(offset), data);
+
+		// VGA I/O byte.
+		if(offset == MCABus::POS::OPTION_SELECT_DATA_1)
+			dynamic_cast<mca16_planar_vga_device *>(m_planar_vga->get_card_device())->sleep_w(data);
+		return;
+	}
+
+	if(BIT(m_system_board_setup, 7) == 0)
+	{
+		if(!machine().side_effects_disabled()) LOGPOS("%s: writing planar POS register %s: %02X\n", FUNCNAME, posreg_name(offset), data);
+
+		m_planar->planar_pos_w(offset, data);
+		return;
+	}
+
+	if(m_planar->get_mca_slot(m_adapter_setup & 7) != NULL)
+	{
+		auto card = dynamic_cast<device_mca16_card_interface *>(m_planar->get_mca_slot(m_adapter_setup & 7)->get_card_device());
+
+		if(card)
+		{
+			LOGPOS("%s: writing adapter %d POS register %s: %02X\n", FUNCNAME, (m_adapter_setup & 7) + 1, posreg_name(offset), data);
+			card->pos_w(offset, data);
```

Comment:
More of a suggestion than anything, but this looks an `address_space` to me, with the virtual `get_card_id` fn logic that can be abstracted in the parent virtual class and children only sets up those values thru a an init setter, like the correlated  fn for PCI arch https://github.com/mamedev/mame/blob/a9519ebcd8cbf9cd48454828cac96c3d5b735376/src/devices/machine/pci.h#L15